csrc/python_frontend/fusion_definition.cpp

```diff
@@ -130,14 +131,38 @@ void FusionDefinition::setupSchedule(const at::ArrayRef<c10::IValue>& inputs) {
 void FusionDefinition::finalizeSchedule(
     const at::ArrayRef<c10::IValue>& inputs) {
   FUSER_PERF_SCOPE("FusionDefinition::finalizeSchedule");
+#ifdef NVFUSER_DISTRIBUTED
+  // TODO: remove when multidevice executor integration is done natively
+  Fusion* fusion = user_sched_->schedule.get();
+  std::vector<Val*> fusion_inputs = InputsOf::outputs(fusion->outputs());
+  std::vector<Val*> vals = DependencyCheck::getAllValsBetween(
+      {fusion_inputs.begin(), fusion_inputs.end()}, fusion->outputs());
+  static Communicator* comm = new Communicator();
+  if (std::any_of(vals.begin(), vals.end(), [](Val* v) {
+        return v->isA<TensorView>() && v->as<TensorView>()->hasDeviceMesh();
+      })) {
+    multidevice_executor_ = std::make_unique<MultiDeviceExecutor>(
+        std::make_unique<Fusion>(*fusion), *comm);
+  }
+
+  FusionGuard::setCurFusion(prev_fusion_);
+  prev_fusion_ = nullptr;
+  if (!multidevice_executor_) {
```

Comment:
```suggestion
  if (multidevice_executor_ == nullptr) {
```

Reply:
I guess you don't like the duplication on Line 158 either :) I think, if we declare (but not define) MultiDeviceExecutor when `!NVFUSER_DISTRIBUTED`, we should be able to use std::unique_ptr<MultiDeviceExecutor> as long as we don't construct the executor. 

That being said, once we hopefully localize NVFUSER_DISTRIBUTED to Communicator, we'll stop worry about all these.

Reply:
Good call. Let me try that

Reply:
Capture off line conversation.
An incomplete type `MultiDeviceExecutor` won't work here with `std::unique_ptr<MultiDeviceExecutor>`. `~unique_ptr` requires calls `sizeof` on the instance, and we do not have a definition of it at all :cry: 

We'll live with the duplicated code.